presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -126,11 +144,147 @@ public PlanNode visitTableScan(TableScanNode node, RewriteContext<Void> context)
+                        .collect(toList());
+                return combinePredicates(child.getType().flip(), negatedPredicates);
+            }
+            else if (node.getValue() instanceof ComparisonExpression && ((ComparisonExpression) node.getValue()).getType() != IS_DISTINCT_FROM) {
+                ComparisonExpression child = (ComparisonExpression) node.getValue();
+                return new ComparisonExpression(
+                        child.getType().negate(),
+                        treeRewriter.rewrite(child.getLeft(), context),
+                        treeRewriter.rewrite(child.getRight(), context));
+            }
+            else if (node.getValue() instanceof NotExpression) {
+                NotExpression child = (NotExpression) node.getValue();
+                return treeRewriter.rewrite(child.getValue(), context);
+            }
+
+            return new NotExpression(treeRewriter.rewrite(node.getValue(), context));
+        }
+    }
+
+    private enum NodeContext
+    {
+        ROOT_NODE,
+        NOT_ROOT_NODE;
+
+        boolean isRootNode()
+        {
+            return this == ROOT_NODE;
+        }
+    }
+
+    private static class ExtractCommonPredicatesExpressionRewriter
+            extends ExpressionRewriter<NodeContext>
+    {
+        @Override
+        public Expression rewriteExpression(Expression node, NodeContext context, ExpressionTreeRewriter<NodeContext> treeRewriter)
+        {
+            if (context.isRootNode()) {
+                return treeRewriter.rewrite(node, NodeContext.NOT_ROOT_NODE);
+            }
+
+            return null;
+        }
+
+        @Override
+        public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, NodeContext context, ExpressionTreeRewriter<NodeContext> treeRewriter)
+        {
+            List<Expression> predicates = extractPredicates(node.getType(), node).stream()
+                    .map(expression -> treeRewriter.rewrite(expression, NodeContext.NOT_ROOT_NODE))
+                    .collect(toList());
+
+            List<List<Expression>> subPredicates = getSubPredicates(predicates);
+            List<Expression> leafPredicates = getLeafPredicates(predicates);
+
+            List<Set<Expression>> deterministicSubPredicates = subPredicates.stream()
+                    .map(this::filterDeterministicPredicates)
+                    .collect(toList());
+
+            Set<Expression> commonPredicates = new HashSet<>(deterministicSubPredicates.stream()
+                    .reduce(Sets::intersection)
+                    .orElse(emptySet()));
+
+            List<List<Expression>> uncorrelatedSubPredicates = subPredicates.stream()
+                    .map(predicateList -> removeAll(predicateList, commonPredicates))
+                    .collect(toList());
+
+            // Do not simplify if a tree structure wouldn't change (apart from simple duplicates removal)
+            if (commonPredicates.isEmpty() || subPredicates.size() == 1) {
```

Comment:
I believe if you have handled all the base cases correctly, you don't need to handle these cases explicitly. The transformations should produce the same result regardless. For simplicity, lets remove this check and make sure everything works as expected.